src/pulsar_client.erl

```diff
@@ -14,456 +14,172 @@
-        {ok, {Sock1, Opts1}} ->
-            pulsar_socket:send_lookup_topic_packet(Sock1, Topic, RequestId, Opts1),
-            {noreply, next_request_id(State#state{
-                requests = maps:put(RequestId, {From, Topic}, Reqs),
-                sock = Sock1,
-                opts = Opts1
-            })}
-    end;
-handle_call(get_status, From, State = #state{sock = undefined, opts = Opts, servers = Servers}) ->
-    case get_alive_sock_opts(Servers, undefined, Opts) of
-        {error, Reason} ->
-            log_error("get_status from pulsar failed: ~p", [Reason]),
-            start_reconnect_timer(),
-            {reply, false, State};
-        {ok, {Sock, Opts1}} ->
-            {reply, not is_pong_longtime_no_received(),
-                State#state{from = From, sock = Sock, opts = Opts1}}
-    end;
-handle_call(get_status, _From, State) ->
-    {reply, not is_pong_longtime_no_received(), State};
-handle_call(get_alive_pulsar_url, From, State = #state{sock = Sock, opts = Opts, servers = Servers}) ->
-    case get_alive_sock_opts(Servers, Sock, Opts) of
-        {error, _Reason} ->
-            start_reconnect_timer(),
-            {reply, {error, no_servers_available}, State};
-        {ok, {Sock1, Opts1}} ->
-            {reply, pulsar_socket:get_pulsar_uri(Sock1, Opts),
-                State#state{from = From, sock = Sock1, opts = Opts1}}
-    end;
-handle_call(_Req, _From, State) ->
-    {reply, ok, State, hibernate}.
+get_alive_pulsar_url(ClientId) ->
+    maybe
+        %% TODO: return all urls
+        {ok, WorkerPid} ?= pick_random_worker(ClientId),
+        pulsar_client_worker:get_alive_pulsar_url(WorkerPid)
+    end.
+
+register_worker(ClientId, WorkerPid, ServiceURL0) ->
+    ServiceURL = bin(ServiceURL0),
+    gen_server:call(ClientId, #register_worker{worker_pid = WorkerPid, url = ServiceURL}, infinity).
+
+unregister_worker_async(ClientId, ServiceURL0) ->
+    ServiceURL = bin(ServiceURL0),
+    gen_server:cast(ClientId, #unregister_worker{url = ServiceURL}).
 
+%%--------------------------------------------------------------------
+%% `gen_server' API
+%%--------------------------------------------------------------------
 
-handle_cast({lookup_topic_async, From, PartitionTopic}, State) ->
-    %% re-use the same logic as the call, as the process of looking up
-    %% a topic is itself async in the gen_server:call.
-    self() ! {'$gen_call', From, {lookup_topic, PartitionTopic}},
+init([ClientId, _Opts]) ->
+    process_flag(trap_exit, true),
+    persistent_term:put(?PT_WORKERS_KEY(ClientId), #{}),
+    State = #{client_id => ClientId},
+    {ok, State}.
+
+handle_call(#register_worker{worker_pid = WorkerPid, url = URL}, _From, State) ->
+    #{client_id := ClientId} = State,
+    handle_register_worker(ClientId, WorkerPid, URL),
+    {reply, ok, State};
+handle_call(_Req, _From, State) ->
+    {reply, {error, unknown_call}, State, hibernate}.
+
+handle_cast(#unregister_worker{url = URL}, State) ->
```

Comment:
This process should monitor the workers, so that they could be forcefully killed and cleaned up.